File: reactor/java-reactor/src/main/java/com/example/reactor/javareactor/global/dto/ErrorResponse.java
package com.example.reactor.javareactor.global.dto;


import com.example.reactor.javareactor.global.exception.ServerException;
import lombok.Builder;

@Builder
public record ErrorResponse(
    int code,
    String message
) {
    public static ErrorResponse from(Throwable exception) {
        boolean isServerException = exception instanceof ServerException;

        return ErrorResponse.builder()
                            .code(isServerException ? ((ServerException) exception).getCode() : 500)
                            .message(isServerException ? ((ServerException) exception).getMessage() : "Internal Server Error")
                            .build();
    }
}

File: reactor/kotlin-reactor/src/main/kotlin/com/example/reactor/kotlinreactor/global/dto/ErrorResponse.kt
package com.example.reactor.kotlinreactor.global.dto

import com.example.reactor.kotlinreactor.global.exception.ServerException

data class ErrorResponse(
    val code: Int,
    val message: String
) {
    companion object {
        operator fun invoke(exception: Throwable): ErrorResponse =
            when (exception) {
                is ServerException -> ErrorResponse(code = exception.code, message = exception.message)
                else -> ErrorResponse(code = 500, message = "Internal Server Error")
            }
    }
}